### PR TITLE
Improved Cash Selection interface

### PR DESCRIPTION
**Pull request by josecoll**
Cash selection refactoring such that 3d party DB providers are only required to implement Coin Selection SQL logic (and not duplicate the locking and retry logic).

This work is a pre-requisite to enable approval of [Experimental support for PostgreSQL ](https://github.com/corda/corda/pull/1525).

### CODE REVIEW

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
+
+/**
+ * Pluggable interface to allow for different cash selection provider implementations
+ * Default implementation [CashSelectionH2Impl] uses H2 database and a custom function within H2 to perform aggregation.
+ * Custom implementations must implement this interface and declare their implementation in
+ * META-INF/services/net.corda.contracts.asset.CashSelection
+ */
+abstract class CashSelection {
```
Comment (vkolomeyko) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: How about calling it `AbstractCashSelection` ?

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                } ?: throw ClassNotFoundException("\nUnable to load compatible cash selection algorithm implementation for JDBC driver ($_metadata)." +
+                        "\nPlease specify an implementation in META-INF/services/net.corda.finance.contracts.asset.CashSelection")
+            }.invoke()
+        }
+
+        val log = loggerFor<CashSelection>()
+    }
+
+    // coin selection retry loop counter, sleep (msecs) and lock for selecting states
+    private val MAX_RETRIES = 5
```
Comment (vkolomeyko) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: Would it not be a good idea to have constants configurable?

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt [hunk 3]
```diff
@@ -33,118 +19,33 @@ class CashSelectionH2Impl : CashSelection {
...
+    override fun executeQuery(statement: Statement, amount: Amount<Currency>, lockId: UUID, notary: Party?,
+                              issuerKeysStr: String?, issuerRefsStr: String?) : ResultSet {
+        statement.execute("CALL SET(@t, 0);")
+
+        val selectJoin = """
+        SELECT vs.transaction_id, vs.output_index, vs.contract_state, ccs.pennies, SET(@t, ifnull(@t,0)+ccs.pennies) total_pennies, vs.lock_id
+        FROM vault_states AS vs, contract_cash_states AS ccs
+        WHERE vs.transaction_id = ccs.transaction_id AND vs.output_index = ccs.output_index
+        AND vs.state_status = 0
+        AND ccs.ccy_code = '${amount.token}' and @t < ${amount.quantity}
```
Comment (vkolomeyko) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt]: This code is prone to SQL injection. Would it not be a better idea to use `PreparedStatement`?

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        fun getInstance(metadata: () -> java.sql.DatabaseMetaData): CashSelection {
+            return instance.get() ?: {
+                val _metadata = metadata()
+                val cashSelectionAlgos = ServiceLoader.load(CashSelection::class.java).toList()
+                val cashSelectionAlgo = cashSelectionAlgos.firstOrNull { it.isCompatible(_metadata) }
+                cashSelectionAlgo?.let {
+                    instance.set(cashSelectionAlgo)
+                    cashSelectionAlgo
+                } ?: throw ClassNotFoundException("\nUnable to load compatible cash selection algorithm implementation for JDBC driver ($_metadata)." +
+                        "\nPlease specify an implementation in META-INF/services/net.corda.finance.contracts.asset.CashSelection")
```
Comment (vkolomeyko) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: In current form should be: `META-INF\services\net.corda.finance.contracts.asset.cash.selection.CashSelection`
That said would not be better to use `META-INF/services/${CashSelection::class.java}` ?

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                    }
+                    log.trace("Coin selection requested $amount but retrieved $totalPennies pennies with state refs: ${stateAndRefs.map { it.ref }}")
+                    // retry as more states may become available
+                } catch (e: SQLException) {
+                    log.error("""Failed retrieving unconsumed states for: amount [$amount], onlyFromIssuerParties [$onlyFromIssuerParties], notary [$notary], lockId [$lockId]
+                        $e.
+                    """)
+                } catch (e: StatesNotAvailableException) { // Should never happen with single threaded state machine
+                    stateAndRefs.clear()
+                    log.warn(e.message)
```
Comment (vkolomeyko) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: Would it make sense to log the full exception stack trace?

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+                        // With the current single threaded state machine available states are guaranteed to lock.
+                        // TODO However, we will have to revisit these methods in the future multi-threaded.
+                        services.vaultService.softLockReserve(lockId, (stateAndRefs.map { it.ref }).toNonEmptySet())
+                        return stateAndRefs
+                    }
+                    log.trace("Coin selection requested $amount but retrieved $totalPennies pennies with state refs: ${stateAndRefs.map { it.ref }}")
+                    // retry as more states may become available
+                } catch (e: SQLException) {
+                    log.error("""Failed retrieving unconsumed states for: amount [$amount], onlyFromIssuerParties [$onlyFromIssuerParties], notary [$notary], lockId [$lockId]
```
Comment (vkolomeyko) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: Ditto for full exception stack trace.

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt (hunk 1 above)
Comment (josecoll) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: Only problem is we end up with a nasty META-INF.services filename:
`net.corda.finance.contracts.asset.cash.selection.AbstractCashSelection`
Does this seem reasonable?

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt (hunk 2 above)
Comment (josecoll) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: Yes - I've added a TODO to make parameters configurable when we get CorDapp configuration (WIP).

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt (hunk 5 above)
Comment (josecoll) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: No ,as this is just a warning message - before next retry.

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt (hunk 6 above)
Comment (josecoll) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: do you mean `${e.getStackTraceAsString()}.` ?

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+                    // the softLockReserve update will detect whether we try to lock states locked by others
+                    val rs = executeQuery(statement, amount, lockId, notary, issuerKeysStr, issuerRefsStr)
+                    stateAndRefs.clear()
+
+                    var totalPennies = 0L
+                    while (rs.next()) {
+                        val txHash = SecureHash.parse(rs.getString(1))
+                        val index = rs.getInt(2)
+                        val stateRef = StateRef(txHash, index)
+                        val state = rs.getBytes(3).deserialize<TransactionState<Cash.State>>(context = SerializationDefaults.STORAGE_CONTEXT)
```
Comment (tomtau) [inline finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: @josecoll @vkolomeyko In Corda 1.0 and master, `VaultSchema` has `contractState` annotated with `@Lob` -- that by default ends up stored in the `pg_largeobject` table in PostgreSQL. This column then holds a reference OID to `pg_largeobject` -- so, if you call `getBytes`, you'll get OIDs instead objects (so, this line would throw a deserialisation exception with PostgreSQL). For actual objects, one needs to call `getBlob`: https://github.com/corda/corda/pull/1525/files#diff-9812d8cb71bc627256a3b382ebc5cc92R129

I guess there are two options

1. use ` rs.getBlob(3)` instead and add a corresponding extension method for deserializing it
2. add / change annotations or configurations: https://stackoverflow.com/questions/3677380/proper-hibernate-annotation-for-byte

What do you think?

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt (hunk 3 above)
Comment (josecoll) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt]: Good point - I will raise a separate PR to address this.

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt (hunk 6 above)
Comment (vkolomeyko) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: I was more thinking of logging as: `log.error("Long message...", e)`

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt (hunk 3 above)
Comment (josecoll) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt]: After some experimentation it has come to light that we cannot use a PreparedStatement for H2:

`[INFO ] 11:17:11,459 [main] (CashSelectionH2Impl.kt:104) selection.CashSelectionH2Impl.unconsumedCashStatesForSpending - HikariProxyPreparedStatement@1497215635 wrapping prep53: 
                    SELECT vs.transaction_id, vs.output_index, vs.contract_state, ccs.pennies, SET(@t, ifnull(@t,0)+ccs.pennies) total_pennies, vs.lock_id
                    FROM vault_states AS vs, contract_cash_states AS ccs
                    WHERE vs.transaction_id = ccs.transaction_id AND vs.output_index = ccs.output_index
                    AND vs.state_status = 0
                    AND ccs.ccy_code = ? and @t < ?
                    AND (vs.lock_id = ? OR vs.lock_id is null)
                     AND vs.notary_name = ? {1: 'USD', 2: 10000, 3: 'ba09bf09-57c1-4eff-a344-b618b4328360', 4: 'C=CH,L=Zurich,O=Notary Service'}
[ERROR] 11:17:11,460 [main] (CashSelectionH2Impl.kt:156) selection.CashSelectionH2Impl.unconsumedCashStatesForSpending - Failed retrieving unconsumed states for: amount [100.00 USD], onlyFromIssuerParties [[]], notary [C=CH,L=Zurich,O=Notary Service], lockId [ba09bf09-57c1-4eff-a344-b618b4328360]
                        org.h2.jdbc.JdbcSQLException: This method is not allowed for a prepared statement; use a regular statement instead. [90130-194].`

Also see http://www.h2database.com/javadoc/org/h2/jdbc/JdbcPreparedStatement.html 
Will investigate further.

Comment (josecoll) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt]: Bizarrely, it is incorrect to use a `PreparedStatement` according to H2 error code docs:
https://www.h2database.com/javadoc/org/h2/api/ErrorCode.html#c90130

Comment (vkolomeyko) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt]: I do not have recent experience with H2's prepared statements, but this chapter: http://www.h2database.com/html/advanced.html#sql_injection
Seems to suggest that it should be possible for our use case.

Comment (josecoll) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelectionH2Impl.kt]: Thanks Victor. Finally got it working ;-)

File: finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt (hunk 7 above)
Comment (josecoll) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: thanks @tomtau. Addressed with option 1) as per your PostgreSQL code submission.

Comment (tomtau) [reply finance/src/main/kotlin/net/corda/finance/contracts/asset/cash/selection/CashSelection.kt]: Thanks a lot!

File: core/src/main/kotlin/net/corda/core/serialization/SerializationAPI.kt [hunk 8]
```diff
@@ -186,6 +187,11 @@ inline fun <reified T : Any> SerializedBytes<T>.deserialize(serializationFactory
 inline fun <reified T : Any> ByteArray.deserialize(serializationFactory: SerializationFactory = SerializationFactory.defaultFactory, context: SerializationContext = serializationFactory.defaultContext): T = this.sequence().deserialize(serializationFactory, context)
 
 /**
+ * Convenience extension method for deserializing a JDBC Blob, utilising the defaults.
+ */
+inline fun <reified T : Any> Blob.deserialize(serializationFactory: SerializationFactory = SerializationFactory.defaultFactory, context: SerializationContext = serializationFactory.defaultContext): T = this.getBytes(1, this.length().toInt()).deserialize(serializationFactory, context)
```
Comment (tomtau) [inline core/src/main/kotlin/net/corda/core/serialization/SerializationAPI.kt]: @josecoll @vkolomeyko btw, where is the blob freed?

Comment (josecoll) [reply core/src/main/kotlin/net/corda/core/serialization/SerializationAPI.kt]: It will be automatically disposed of when the result set is closed.

**Pull request merged**